File: rust/rust-log/src/function.rs
// 由关键字```fn```声明
// 没有可选参数
// 没有默认参数值
// 可利用**元组**可以返回多个不同类型的值

// 没有 `return` 语句则使用最后一条语句的结果作为返回值

// 无参数，无返回值
fn func_void(){
    println!("func_void")
}

// 有参数，多个返回值(返回元组)
fn func_args(a:i32, b:bool, s: String) -> (i32, bool) {
    println!("{},{},{}", a, b, s);
    return (a, b);
}

pub fn func_run() {
    func_void();                      // 结果：func_void
    let (args1, args2) = func_args(1, true, String::from("函数")); // 结果：1,true,函数
    println!("{},{}", args1, args2); // 结果：1,true
    
    // 匿名(闭包)函数 lambda
    let sum = |a: i32, b: i32| -> i32 {
        a+b
    };
    // 相当于 fn sum(a: i32, b: i32) -> i32 {}

    println!("匿名函数:{}", sum(1,3)); // 结果：4
    
}

File: z_other/python/py-log/basic/function.py
# 由关键字```def```声明
# 有默认参数值
# 有可变长参数(不定长参数)

# 函数的参数传递是通过对象引用来完成的，在内部修改是否影响外部数据要看参数是否是可变类型(list dict)

def func_run():
    func_void()  # 结果： void
    func_names('a', 'b', 'c')  # 结果： ('a', 'b', 'c') 3

    t: int = func_args(1)  # 结果： 1 False str...
    print(t)  # 结果： 1

    dicts = {'name': 'laixhe', 'age': 18}
    func_key(**dicts)  # 结果： {'name': 'laixhe', 'age': 18}
    pass


# 无参数，无返回值
def func_void():
    print('void')
    pass


# 可变长参数(不定长参数)，参数会以元组(tuple)的形式导入
def func_names(*s: str):
    print(s, len(s))
    pass


# 有参数和默认值，有返回值
def func_args(a: int, b: bool = False, s: str = 'str...') -> int:
    print(a, b, s)
    return a


# 关键字可变参数(不定长参数)，参数会以字典(dict)的形式导入
def func_key(**kwargs: dict):
    print(kwargs)
    pass

